core/src/sumeragi/main_loop.rs

```diff
@@ -138,48 +146,45 @@ impl Sumeragi {
     fn receive_network_packet(
         &self,
         view_change_proof_chain: &mut ProofChain,
-        control_message_in_a_row_counter: &mut usize,
-    ) -> Option<Message> {
+    ) -> (Option<BlockMessage>, bool) {
         const MAX_CONTROL_MSG_IN_A_ROW: usize = 25;
 
-        if *control_message_in_a_row_counter < MAX_CONTROL_MSG_IN_A_ROW {
-            *control_message_in_a_row_counter += 1;
-            self.control_message_receiver
+        let mut should_sleep = true;
+        for _ in 0..MAX_CONTROL_MSG_IN_A_ROW {
+            if let Ok(msg) = self.control_message_receiver
                 .try_recv()
                 .map_err(|recv_error| {
                     assert!(
                         recv_error != mpsc::TryRecvError::Disconnected,
                         "Sumeragi control message pump disconnected. This is not a recoverable error."
                     )
-                })
-                .ok()
-                .map(std::convert::Into::into)
-        } else {
-            None
-        }.or_else(|| {
-            *control_message_in_a_row_counter = 0;
-            self
-                .message_receiver
-                .try_recv()
-                .map_err(|recv_error| {
-                    assert!(
-                        recv_error != mpsc::TryRecvError::Disconnected,
-                        "Sumeragi message pump disconnected. This is not a recoverable error."
-                    )
-                })
-                .ok()
-        })
-        .and_then(|packet : MessagePacket| {
-            if let Err(error) = view_change_proof_chain.merge(
-                packet.view_change_proofs,
-                &self.current_topology.ordered_peers,
-                self.current_topology.max_faults(),
-                self.wsv.latest_block_hash(),
-            ) {
-                trace!(%error, "Failed to add proofs into view change proof chain")
+                }) {
+                should_sleep = false;
+                if let Err(error) = view_change_proof_chain.merge(
+                    msg.view_change_proofs,
+                    &self.current_topology.ordered_peers,
+                    self.current_topology.max_faults(),
+                    self.wsv.latest_block_hash(),
+                ) {
+                    trace!(%error, "Failed to add proofs into view change proof chain")
+                }
+            } else {
+                break;
             }
-            packet.message
-        })
+        }
+
+        let msg = self
+            .message_receiver
+            .try_recv()
+            .map_err(|recv_error| {
+                assert!(
+                    recv_error != mpsc::TryRecvError::Disconnected,
+                    "Sumeragi message pump disconnected. This is not a recoverable error."
+                )
+            })
+            .ok();
+        should_sleep &= msg.is_none();
+        (msg, should_sleep)
     }
```

Comment:
depending on the previously received control flow messages, is it possible that some of the `BlockMessage`s should just be rejected?

Reply:
Yes, e.g. if enough proof received view change should be updated making block with previous view change irrelevant.

Reply:
so we should also listen for block messages here in a loop while invalid block messages are observed